File: draft_helper.py
#!/usr/bin/env python3
"""
Calcutta Draft Helper — Youche CC

Reads analysis-results.json and shows:
  1. Each flight ranked by true talent (avg differential)
  2. Snake draft order with top B/C picks for each A player

Usage: python3 draft_helper.py
"""

import json
import sys
from pathlib import Path

INPUT = Path(__file__).parent / "analysis-results.json"

def trend_icon(trend: str) -> str:
    return {"improving": "📈", "declining": "📉"}.get(trend, "➡️ ")

def flag(r: dict) -> str:
    if r.get("error"):          return "❓ NO DATA "
    s = r["sandbagger_score"]
    if s >= 7:                  return "🚨 SANDBAG "
    if s >= 4:                  return "⚠️  WATCH   "
    if r["trend"] == "improving": return "📈 HOT     "
    if r["trend"] == "declining": return "📉 COLD    "
    return                             "✅ CLEAN   "

def print_flight(players: list[dict], flight: str):
    ranked = sorted(players, key=lambda p: p["avg_differential"])
    print(f"\n{'═'*78}")
    print(f"  Flight {flight} — Ranked by True Talent  (avg differential, lower = better)")
    print(f"{'═'*78}")
    print(f"  {'#':<4}{'Name':<26}{'HCP':<7}{'Avg Diff':<10}{'Delta':<9}{'Vol':<8}{'Rds':<5}Status")
    print("  " + "─"*76)
    for i, p in enumerate(ranked, 1):
        delta = f"+{p['diff_vs_hcp']}" if p["diff_vs_hcp"] > 0 else str(p["diff_vs_hcp"])
        note  = f" ({p['rounds_fetched']}rds)" if p["rounds_fetched"] < 5 else ""
        print(f"  {i:<4}{p['name']:<26}{p['posted_hcp']:<7}{p['avg_differential']:<10}"
              f"{delta:<9}{p['volatility']:<8}{p['rounds_fetched']:<5}{flag(p)}{note}")

def main():
    if not INPUT.exists():
        print(f"\n❌  Run analyze.py first to generate {INPUT}")
        sys.exit(1)

    all_players: list[dict] = json.loads(INPUT.read_text())

    by_flight = {f: [p for p in all_players if p["flight"] == f] for f in "ABC"}

    print("\n🏌️  CALCUTTA DRAFT HELPER — Youche CC")
    print(f"    {len(all_players)} players | "
          f"A:{len(by_flight['A'])}  B:{len(by_flight['B'])}  C:{len(by_flight['C'])}")

    for flight in "ABC":
        print_flight(by_flight[flight], flight)

    # Snake draft — A players sorted highest HCP first (highest picks first in round 1)
    a_by_hcp  = sorted(by_flight["A"], key=lambda p: -p["posted_hcp"])
    b_ranked  = sorted(by_flight["B"], key=lambda p: p["avg_differential"])
    c_ranked  = sorted(by_flight["C"], key=lambda p: p["avg_differential"])

    print(f"\n\n{'═'*78}")
    print("  SNAKE DRAFT ORDER  (A players, highest HCP picks first)")
    print("  Shows top available B & C picks by true talent (avg differential)")
    print("  🚨 = sandbagger score ≥ 5  |  delta = posted HCP − avg diff (+= plays better)")
    print(f"{'═'*78}\n")

    for i, a in enumerate(a_by_hcp, 1):
        top_b = b_ranked[:3]
        top_c = c_ranked[:3]
        delta = f"+{a['diff_vs_hcp']}" if a["diff_vs_hcp"] > 0 else str(a["diff_vs_hcp"])

        print(f"  Pick {i:<3} {a['name']}  (HCP {a['posted_hcp']}, Avg Diff {a['avg_differential']}, Δ{delta})")

        def fmt(p: dict) -> str:
            warn = " 🚨" if p["sandbagger_score"] >= 5 else ""
            return f"{p['name']} [{p['avg_differential']}{warn}]"

        if top_b:
            print(f"          B picks: {' | '.join(fmt(p) for p in top_b)}")
        if top_c:
            print(f"          C picks: {' | '.join(fmt(p) for p in top_c)}")
        print()

    print("  ⚠️   Note: this does NOT remove picked players as you draft.")
    print("       Cross them off manually and re-run for live updates.\n")

if __name__ == "__main__":
    main()

File: draft-helper.ts
/**
 * Calcutta Draft Helper — Youche CC
 *
 * Reads analysis-results.json and presents a snake-draft order view
 * showing each A player's best available B and C picks, ranked by
 * "true talent" (avg differential) rather than posted handicap.
 *
 * Usage: npm run draft
 */

import fs from 'fs';
import path from 'path';

const INPUT = path.resolve('./analysis-results.json');

interface PlayerAnalysis {
  name: string;
  ghinId: string;
  postedHandicap: number;
  flight: 'A' | 'B' | 'C';
  roundsFetched: number;
  avgDifferential: number;
  diffVsHandicap: number;
  sandbaggerScore: number;
  volatility: number;
  trend: 'improving' | 'declining' | 'stable';
  recentDifferentials: number[];
  error?: string;
}

function trendEmoji(trend: string) {
  if (trend === 'improving') return '📈';
  if (trend === 'declining') return '📉';
  return '➡️ ';
}

function flag(p: PlayerAnalysis): string {
  if (p.error) return '❓';
  if (p.sandbaggerScore >= 7) return '🚨 SANDBAG';
  if (p.sandbaggerScore >= 4) return '⚠️  WATCH  ';
  if (p.trend === 'improving') return '📈 HOT    ';
  if (p.trend === 'declining') return '📉 COLD   ';
  return '✅ CLEAN  ';
}

function printRankedFlight(players: PlayerAnalysis[], flight: string) {
  console.log(`\n${'═'.repeat(75)}`);
  console.log(`  Flight ${flight} — Ranked by True Talent (avg differential, low = better)`);
  console.log('═'.repeat(75));
  console.log(`  ${'#'.padEnd(4)}${'Name'.padEnd(26)}${'HCP'.padEnd(7)}${'Avg Diff'.padEnd(10)}${'Delta'.padEnd(9)}${'Volatility'.padEnd(12)}Status`);
  console.log('  ' + '─'.repeat(73));

  players.forEach((p, i) => {
    const delta = p.diffVsHandicap > 0 ? `+${p.diffVsHandicap}` : `${p.diffVsHandicap}`;
    const rounds = p.roundsFetched < 5 ? ` (${p.roundsFetched}rds)` : '';
    console.log(
      `  ${String(i + 1).padEnd(4)}${p.name.padEnd(26)}${String(p.postedHandicap).padEnd(7)}${String(p.avgDifferential).padEnd(10)}${delta.padEnd(9)}${String(p.volatility).padEnd(12)}${flag(p)}${rounds}`
    );
  });
}

function main() {
  if (!fs.existsSync(INPUT)) {
    console.error(`\n❌  Run "npm run analyze" first to generate ${INPUT}`);
    process.exit(1);
  }

  const all: PlayerAnalysis[] = JSON.parse(fs.readFileSync(INPUT, 'utf8'));

  // Sort each flight by avg differential ascending (lower = plays better = picked first)
  const byFlight = (flight: 'A' | 'B' | 'C') =>
    all
      .filter(p => p.flight === flight)
      .sort((a, b) => a.avgDifferential - b.avgDifferential);

  const aPlayers = byFlight('A');
  const bPlayers = byFlight('B');
  const cPlayers = byFlight('C');

  console.log('\n🏌️  CALCUTTA DRAFT HELPER — Youche CC');
  console.log(`    Analysis based on last ${all[0]?.recentDifferentials?.length ?? '?'} rounds per player\n`);

  printRankedFlight(aPlayers, 'A');
  printRankedFlight(bPlayers, 'B');
  printRankedFlight(cPlayers, 'C');

  // Snake draft simulation — A players draft in reverse handicap order (lowest HCP picks last in round 1)
  console.log(`\n${'═'.repeat(75)}`);
  console.log('  SNAKE DRAFT ORDER (A players, highest HCP picks first)');
  console.log('═'.repeat(75));
  console.log('  Each A player\'s top 3 available B and C picks by true talent:\n');

  const aByHandicap = [...all.filter(p => p.flight === 'A')]
    .sort((a, b) => b.postedHandicap - a.postedHandicap); // highest HCP picks first

  const remainingB = [...bPlayers];
  const remainingC = [...cPlayers];

  aByHandicap.forEach((a, i) => {
    const pickNum = i + 1;
    const topB = remainingB.slice(0, 3);
    const topC = remainingC.slice(0, 3);

    console.log(`  Pick ${String(pickNum).padEnd(3)} ${a.name} (HCP ${a.postedHandicap}, Avg Diff ${a.avgDifferential})`);

    if (topB.length) {
      console.log(`         B options: ${topB.map(p => `${p.name} [${p.avgDifferential}${p.sandbaggerScore >= 5 ? ' 🚨' : ''}]`).join('  |  ')}`);
    }
    if (topC.length) {
      console.log(`         C options: ${topC.map(p => `${p.name} [${p.avgDifferential}${p.sandbaggerScore >= 5 ? ' 🚨' : ''}]`).join('  |  ')}`);
    }
    console.log('');
  });

  console.log('\n  🚨 = sandbag score ≥ 5/10 — consider carefully before drafting');
  console.log('  Delta = posted HCP minus avg differential (+= plays BETTER than handicap)\n');
}

main();
